File: competition/01.周赛/331.周赛场/02.统计范围内的元音字符串数.js
/**
 * @param {string[]} words
 * @param {number[][]} queries
 * @return {number[]}
 */
var vowelStrings = function (words, queries) {
  const sMap = {
    a: true,
    e: true,
    i: true,
    o: true,
    u: true,
  }
  const map2 = {}
  const res = Array.from({ length: queries.length }, () => 0)
  for (let i = 0; i < words.length; i++) {
    const c1 = words[i].charAt(0)
    const c2 = words[i].charAt(words[i].length - 1)
    map2[i] = sMap[c1] && sMap[c2]
  }
  for (let i = 0; i < queries.length; i++) {
    const query = queries[i]
    for (let j = query[0]; j <= query[1]; j++) {
      if (map2[j]) {
        res[i]++
      }
    }
  }
  return res
}

vowelStrings(
  ['aba', 'bcb', 'ece', 'aa', 'e'],
  [
    [0, 2],
    [1, 4],
    [1, 1],
  ]
)

File: competition/01.周赛/331.周赛场/02. 统计范围内的元音字符串数.java
class Solution {
    public int[] vowelStrings(String[] words, int[][] queries) {
  HashSet<Character> set = new HashSet<>();
        set.add('a');
        set.add('e');
        set.add('i');
        set.add('o');
        set.add('u');
        int[] temp = new int[words.length + 1];
        int[] answer = new int[queries.length];
        for (int i = 0; i < words.length; i++) {
            if (set.contains(words[i].charAt(0)) && set.contains(words[i].charAt(words[i].length() - 1))) temp[i + 1] = temp[i] + 1;
            else temp[i + 1] = temp[i];
        }
        for (int i = 0; i < queries.length; i++) {
            int[] query = queries[i];
            answer[i] = temp[query[1] + 1] - temp[query[0]];
        }
        return answer;
    }
}